[Pughal77] ip

Pughal77 opened the pull request:
# Sir Duke

With **Sir Duke**, you would be able to:
- Arrange your tasks and keep track of them
- Be happy

Also it has the following commands:
1. Add
2. Delete
3. Deadline
4. Event
5. List
6. Find
and more is yet to come!

Click [here](https://github.com/Pughal77) to go to my github page! 👍
Here is my `main` method found at `src/main/java/SirDuke.java`:
```
public static void main(String[] args) {
        new SirDuke("data/sirDuke.txt").run();
 }
```
How are you doing?
- [ ] fine
- [ ] this mod is killing me
- [ ] Sir Duke is amazing!

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

Pughal77 committed:
lvl 0 completed

Pughal77 committed:
lvl 1 completed. Created 2 new enum classes, Commands and Std_msgs. Std_msgs is an enum class of Msgs of all the standard messages for my chatbot. Commands is an enum class of Strings (to be used in switch-cases later on).

Pughal77 committed:
deleted commands enum as it is redundant to wrap every string input as commands object

Pughal77 committed:
Created items that handles the string formatting and printitng th relevant Msgs

Pughal77 committed:
Level 3 done. Seeing problems with how to handle commands in later stages

Pughal77 committed:
Level 4 achieved. I think i should create a handler for user cmds and add a new cmd "add" that adds more tasks to Duke.items

Pughal77 committed:
created first test case in input.txt file

Pughal77 committed:
exception handling for emtpy fields for deadline and event

Pughal77 committed:
able to delete tasks

Pughal77 committed:
Add every class to a package

Added every class to a package as per the coding standard

Pughal77 committed:
Change from if-else to switch-case

Instead of using if-else commands and checking if the user input starts
with a command, I created a enum class called Commands and changed the
if-else clause to a switch-case.

Pughal77 committed:
Add comments for every class

I added comments to every class and method as per the java coding
standard

Pughal77 committed:
Small Edit to Readme.md

Pughal77 committed:
Create new Package called Handler

I realised when implementing the parsing of duke.txt and loading the
contacts that the main method of Sir Duke was getting messy. With the
long block of switch-case, I decided to create a new package called the
handler, where I store functions to store how the parser is handled and
commands are going to be handled.

Pughal77 committed:
Add dialogue for empty list

When the user inputs the list command when the list was empty a message
will be displayed to user that no tasks are in the list

Pughal77 committed:
Implement level-7

Now Users are able to save and load their tasks.

Pughal77 committed:
Delete data dir and sirDuke.txt

I would like my remote repository to not have the data directory
and sirduke.txt

Pughal77 committed:
Supports date time

User should input in the format dd-MM-yyyy HH:mm
and it is written to sirduke.txt in the format of MMM d yyyy HH:mm

Pughal77 committed:
Add sirduke.txt to .gitignore

Pughal77 committed:
Merge branch 'master' into branch-level-8

Pughal77 committed:
Merge branch 'branch-level-8'

Pughal77 committed:
Fix Bug

There was a bug where marked tasks would be unmarked when loading data.
This is because when loading the data i did not use the String.equals
method but used the "==" operator instead.

Pughal77 committed:
Rename Items class to TaskList

This is follow more closely to the suggested answer in 2103T website

Pughal77 committed:
Add addTopDivider to Msg class

ReflectiveObsidian commented on src/main/java/dukeexecpetions/DeadlineEmptyException.java:
```diff
@@ -0,0 +1,11 @@
+/** Exception when an empty field is passed for a deadline*/
+public class DeadlineEmptyException extends EmptyFieldException {
+    /**
+     * Asks the user what time the task he
+     * @param userInput
```
Could there be an empty line between description and parameter section?

ReflectiveObsidian commented on src/main/java/dukeexecpetions/EventEmptyException.java:
```diff
@@ -0,0 +1,26 @@
+
+/** Exception when an empty field is passed for a field*/
+public class EventEmptyException extends EmptyFieldException{
+    /** text for when from entry is missing */
+    private static final String missing_from_txt = "from - What time does %s start from?\n";
```
Could this variable name be treated as a constant, and thus be named using SCREAMING_SNAKE_CASE?

ReflectiveObsidian commented on src/main/java/dukeexecpetions/EventEmptyException.java:
```diff
@@ -0,0 +1,26 @@
+     * @param missing_from
+     * @param missing_to
+     */
+    public EventEmptyException(String event, boolean missing_from, boolean missing_to) {
```
Could the variable names use camel case, and reflect their boolean nature?
```suggestion
    public EventEmptyException(String event, boolean isMissingFrom, boolean isMissingTo) {
```

ReflectiveObsidian commented on src/main/java/handler/CommandsHandler.java:
```diff
@@ -0,0 +1,98 @@
+     * @throws EventEmptyException Field to or from is empty
+     * @throws DeadlineEmptyException Field by is empty
+     */
+    public static boolean handleCommand(String userInput, Items items)
```
Could the method name be more specific?
```suggestion
    public static boolean executeCommand(String userInput, Items items)
```

ReflectiveObsidian reviewed (commented):
LGTM, some naming and coding standard feedback.

ararchch commented on src/main/java/msg/StdMsgs.java:
```diff
@@ -0,0 +1,36 @@
+ * His logo is also stored here
+ */
+public enum StdMsgs {
+    WELCOME(new Msg("Greetings Sir/Mdm, I am Sir sirduke, at your service\n" +
```
I like how you have split the long message text into two, but would it be more easy to follow if you split the line before the operator?

ararchch commented on src/main/java/items/Items.java:
```diff
@@ -0,0 +1,112 @@
+     *
+     * @param item
+     */
+    public void add(Task item) throws IOException {
```
I like how you have made sure to put a space before the opening '{' brace, as well as in other appropriate places in the method signature. Have also noticed that you have done this for other methods.

ararchch commented on src/main/java/handler/DataHandler.java:
```diff
@@ -0,0 +1,85 @@
+                // 1. what if another number other than 0 is put in the isDone category?
+                // 2. what if the first character is not a task tag?
+                // 3. how do we detect that the file is of incorrect format?
+                switch (tag) {
```
Would it be more appropriate in terms of convention to have no indentation for `case `clauses?

ararchch commented on src/main/java/items/Items.java:
```diff
@@ -0,0 +1,112 @@
+    /** tracks number of tasks created*/
+    private int item_count = 0;
+    /** ArrayList<task> to store tasks */
+    private ArrayList<Task> item_list = new ArrayList<Task>();
```
Would it enhance code readability if you use plural form to represent the list of items (eg: `items` )?

ararchch commented on src/main/java/handler/DataHandler.java:
```diff
@@ -0,0 +1,85 @@
+        D, // Deadline
+        E // Event
+    }
+    public static void loadData(Items items)
+            throws IOException {
+        File data = new File("data/sirDuke.txt");
+        Scanner parser;
+        try {
+            parser = new Scanner(data);
+        } catch (FileNotFoundException e) {
+            File dir = new File("data");
+            if (dir.exists()) {
+                data.createNewFile();
+            } else {
+                dir.mkdir();
+                data.createNewFile();
+            }
+            parser = new Scanner(data);
+        }
+        while (parser.hasNext()) {
+            // parse through the text
+            String line = parser.nextLine();
+            String[] inputsToCmd;
+            TaskTag tag;
+            Boolean isDone;
+            DateTimeFormatter formatter = DateTimeFormatter.ofPattern("MMM d yyyy HH:mm");
+            try {
+                tag = TaskTag.valueOf(String.valueOf(line.charAt(0)));
```
Could it enhance readability to split this code in the block into logical units separated by one blank line?

ararchch reviewed (commented):
LGTM! Just some nits regarding coding standard that can be improved on.

Pughal77 committed:
Rename Item to task

Pughal77 committed:
Change structure of Duke

Pughal77 committed:
Debug new structure

Pughal77 committed:
Restructure Sir Duke

Instead of having a handler class for my commands and data. I now have
A Parser for handling string inputs(cmdInput and dataFormat) and I have
abstract class called Command that has and abstract method called
execute.

Pughal77 committed:
Merge branch 'branch-A-MoreOOP'

Pughal77 committed:
Merge branch 'branch-add-gradle-support' into branch-A-Gradle

Pughal77 committed:
Add gradle to project

I also added the checkstyle plugin

Pughal77 committed:
Merge branch 'branch-A-Gradle'

Pughal77 committed:
Add more java docs

Pughal77 committed:
Add Find Command

The find command returns a barebones task list that prints tasks
in the same format as a taskList. I also made sure every command
overrides the execute method.

Pughal77 committed:
Add dependicies for javaFx to `build.gradle`

Pughal77 committed:
Do step 3 of tutorial

Pughal77 committed:
Add JUnit tests

Added JUnit tests for my 2 most problematic classes, parser and
taskList. I need to know how to set up JUnit tests without introducing
any new methods for my current classes. I created new get methods to
access the task object in my deadline and event object.

Pughal77 committed:
Merge branch 'branch-JUnit'

Pughal77 committed:
Add Jar using gradle

Pughal77 committed:
Merge branch 'master' into branch-JavaFx-Basics

# Conflicts:
#	build.gradle

resolved

Pughal77 committed:
Complete Stage 4 of our tutorial

I simply added all the files required as stated in the tutorial

Pughal77 committed:
JUnit test for find method in TaskList

RunjiaChen commented on src/main/java/task/Task.java:
```diff
@@ -0,0 +1,85 @@
+    public void unMarkAsDone() {
+        this.isDone = false;
+    }
+
```
Should it be unmarkAsDone isntead?

RunjiaChen commented on src/main/java/parser/Parser.java:
```diff
@@ -0,0 +1,169 @@
+                try {
+                    return new EventCmd(inputs[0],
+                            LocalDateTime.parse(inputs[1].substring(5), formatter),
+                            LocalDateTime.parse(inputs[2].substring(3), formatter));
```
I like how you are cautions of the character limit for each line and inserted line breaks here.

RunjiaChen commented on src/main/java/storage/Storage.java:
```diff
@@ -0,0 +1,91 @@
+        }
+    }
+
+    public void load() {
```
Should this be changed to loadTasks() so as to make the method name more descriptive and explain what it does?

RunjiaChen commented on src/main/java/storage/Storage.java:
```diff
@@ -0,0 +1,91 @@
+        }
+    }
+
+    public void save() {
```
Should this be changed to saveTasks() so as to make the method name more descriptive and explain what it does?

RunjiaChen commented on src/main/java/task/Task.java:
```diff
@@ -0,0 +1,85 @@
+    /** Description of Task */
+    protected String description;
+    /** is true when task isDone */
+    protected boolean isDone;
```
I like how you followed naming conventions and made boolean variables sound more obvious as boolean variables.

RunjiaChen reviewed (commented):
Your code LGTM, with just a few nits to pick.

Pughal77 committed:
Merge branch 'master' of https://github.com/Pughal77/ip into branch-JavaFx-Basics

Pughal77 committed:
Seperate SriDuke and MainWindow

Pughal77 committed:
Change void output to String output

Now all the execute method in command returns a String response

Pughal77 committed:
Edit getResponse Method

Now getResponse returns the response of Sir Duke as well

Pughal77 committed:
GUI setup for SirDuke

The GUI now works

The plan for the future is to brush up a bit on the GUI using scene builder and also to implement wrap text for my dialog box.

Pughal77 committed:
Add assertions

It is crucial for `SirDuke` to have `taskList`, `Ui` and `storage` classes all initialised before it `run` or `runForGUI`.

It is also crucial to check that the `filePath` passed into the `Storage` class is not `""` or `null`

Pughal77 committed:
Merge pull request #2 from Pughal77/branch-A-Assertions

Add assertions

Pughal77 committed:
Shorten methods

The `parseCommand` method in `Parser` is now a shorter method while still only having one layer of abstraction (SLAPping hard)

Pughal77 committed:
Merge pull request #3 from Pughal77/branch-A-CodingQuality

Shorten methods

Pughal77 committed:
Implement Archiving

The are issues where the file is created but nothing is written to the file.

I also need to run storage.save() when all tasks are cleared

Pughal77 committed:
Add `save` method in archiveCmd

Pughal77 committed:
Archive is up!

Now with the command archive you are able to clear all the tasks stored in storage and save them to archive/archived.txt

Pughal77 committed:
Merge pull request #4 from Pughal77/branch-A-Archive

Branch a archive

Pughal77 committed:
GUI update

Upon Starting the application, users are now greeted with a welcome message.

But there are more things that can be done to further improve the GUI
Let's
*Set a background image
*Make the imageView appear circular

Moreover there is a bug that needs to be fixed. Storage.save() should not write the taskList empty message into data/sirDuke.txt

Pughal77 committed:
Update Readme

Pughal77 committed:
Set up for release

Pughal77 committed:
Update README.md

Pughal77 committed:
Bug fixes

I fixed bugs relating to the inconsistent indexing when taking in inputs for `delete`, `mark` and `unmark`.

Pughal77 committed:
Merge branch 'master' of https://github.com/Pughal77/ip

Pughal77 committed:
Bug fixes

There was a bug where if my storage were not loaded with a certain number of tasks the test, `testFind` would fail. Now, it creates 2 tasks, returns them and then deletes them accordingly

Pughal77 committed:
Add Java Docs

Added JavaDocs to relevant methods

Pughal77 committed:
Update Ui.png
